designs/generic-git-provider-gitops.md

```diff
@@ -0,0 +1,205 @@
+```Go
+type Flux struct {
+    // Github is the name of the Git Provider to host the Git repo.
+	Github Github `json:"github,omitempty"`
+
+    // Git is the name of the generic Git Provider to host the Git repository
+    Git Git `json:"git,omitempty"`
+
+    // FluxSystemNamespace scope for this operation. Defaults to flux-system.
+    FluxSystemNamespace string `json:"fluxSystemNamespace,omitempty"`
+    
+    // ClusterConfigPath relative to the repository root, when specified the cluster sync will be scoped to this path.
+    ClusterConfigPath string `json:"clusterConfigPath,omitempty"`
+
+	// Git branch. Defaults to main.
+	// +kubebuilder:default:="main"
+	Branch string `json:"branch,omitempty"`
+}
+```
+
+The `Git` struct contains the requisite fields for authenticating to a generic git repository using basic auth, either password, SSH or both together. Use of `net/url` for the repository URL provides out-of-the-box validations and assurance of structure.
+```Go
+type Git struct {
+    // Username is the user to authenticate to the git repository with.
+    Username string `json:username`
+
+    // Repository URL for the repository to be used with flux. Can be either an SSH or HTTPS url.
+    RepositoryUrl net/url `json:repositoryUrl`
+
+    // Path to a private key file for authenticating to the repository via SSH.
+    PrivateKeyFile string `json:privateKeyFile,omitempty`
+
+    // An environment variable which should contain a user-provided password for authenticating to the git repository. 
+    passwordEnvironmentVariable string `json:passwordEnvironmentVariable,omitempty`
+}
+```
+
+The `Github` struct will now no longer contain the `ClusterConfigPath` and `FluxSystemNamespace` fields, which are not specific to Github. It will contain only the fields required to establish a connection to Github -- Owner, Repository, and wether or not it's a Personal repository.
+```Go
+type Github struct {
+	// Owner is the user or organization name of the Git provider.
+	Owner string `json:"owner"`
+
+	// Repository name.
+	Repository string `json:"repository"`
+
+	// if true, the owner is assumed to be a Git user; otherwise an org.
+	Personal bool `json:"personal,omitempty"`
+}
+```
+
+`GitOpsCOnfig.spec.flux.git.passwordEnvironmentVariable`
+An environment variable which contains the password to be used when authenticating to the git repository
+
+`GitOpsCOnfig.spec.flux.git.privateKeyPath` 
+A path to a private key to be used for authenticating to the Git SSH server
+
+### Security
+#### Getting Credentials
+User may provide a private key, a password, or both.
+The private key will be provided via the path to the private key file.
+The password will be provided via an environment variable; the environment variable will be chosen by the user and later validated to exist by us.
```

Comment:
Can we start with private key first and see if users want username/password instead?